rllib/core/learner/tf/tf_learner.py

```diff
@@ -430,104 +431,36 @@ def build(self) -> None:
             super().build()
 
         if self._enable_tf_function:
-            self._update_fn = tf.function(self._do_update_fn, reduce_retracing=True)
-        else:
-            self._update_fn = self._do_update_fn
+            self._update = tf.function(self._original_update_fn, reduce_retracing=True)
 
     @override(Learner)
-    def update(
-        self,
-        batch: MultiAgentBatch,
-        *,
-        minibatch_size: Optional[int] = None,
-        num_iters: int = 1,
-        reduce_fn: Callable[[ResultDict], ResultDict] = ...,
-    ) -> Mapping[str, Any]:
-        # TODO (Kourosh): The update of learner is vastly differnet than the base
-        #  class. So we need to unify them.
-        missing_module_ids = set(batch.policy_batches.keys()) - set(self._module.keys())
-        if len(missing_module_ids) > 0:
-            raise ValueError(
-                "Batch contains module ids that are not in the learner: "
-                f"{missing_module_ids}"
-            )
-
-        batch_iter = (
-            MiniBatchCyclicIterator
-            if minibatch_size is not None
-            else MiniBatchDummyIterator
-        )
-
-        results = []
-        for minibatch in batch_iter(batch, minibatch_size, num_iters):
-            # TODO (Avnish): converting to tf tensor and then from nested dict back to
-            #  dict will most likely hit us in perf. But let's go with this for now.
-            tensorbatch = self._convert_batch_type(minibatch)
-            update_outs = self._update_fn(tensorbatch)
-            loss = update_outs["loss"]
-            fwd_out = update_outs["fwd_out"]
-            postprocessed_gradients = update_outs["postprocessed_gradients"]
-            result = self.compile_results(batch, fwd_out, loss, postprocessed_gradients)
-            self._check_result(result)
-            results.append(result)
-
-        # Reduce results across all minibatches, if necessary.
-        if len(results) == 1:
-            return results[0]
-        else:
-            if reduce_fn is None:
-                return results
-            return reduce_fn(results)
-
-    def _do_update_fn(
+    def _update(
         self,
-        batch: MultiAgentBatch,
+        batch: NestedDict,
```

Comment:
What is this _ray_trace_ctx? Do we need that? I don't know how it appeared here?

Reply:
There is this super strange interaction between tf-eager-tracing and ray tracing :(

Putting this extra "dummy"'ish arg here does suppress/fix an error. It took me a while to figure this out almost 2 years ago when we cleaned up the eager_tracing_policy stuff.

Reply:
Can we put a comment for it then? Explain the reason? Cite the stack overflow issue, etc?

Reply:
Yeah, will do!